spring-kafka/src/test/java/org/springframework/kafka/core/KafkaTemplateTests.java

```diff
@@ -339,7 +344,66 @@ public void onFailure(Throwable ex) {
 		});
 		assertThat(KafkaTestUtils.getSingleRecord(consumer, INT_KEY_TOPIC)).has(value("foo"));
 		assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
-		pf.createProducer().close();
+		pf.destroy();
+	}
+
+	@SuppressWarnings("unchecked")
+	@Test
+	void testWithCallbackFailure() throws Exception {
+		Producer<Integer, String> producer = mock(Producer.class);
+		willAnswer(inv -> {
+			Callback callback = inv.getArgument(1);
+			callback.onCompletion(null, new RuntimeException("test"));
+			return new SettableListenableFuture<RecordMetadata>();
+		}).given(producer).send(any(), any());
+		ProducerFactory<Integer, String> pf = mock(ProducerFactory.class);
+		given(pf.createProducer()).willReturn(producer);
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf);
+		ListenableFuture<SendResult<Integer, String>> future = template.send("foo", 1, "bar");
+		final CountDownLatch latch = new CountDownLatch(1);
+		final AtomicReference<SendResult<Integer, String>> theResult = new AtomicReference<>();
+		AtomicReference<String> value = new AtomicReference<>();
+		future.addCallback(new KafkaSendCallback<Integer, String>() {
+
+			@Override
+			public void onSuccess(SendResult<Integer, String> result) {
+			}
```

Comment:
How about making this `onSuccess` as a `default` one in the `KafkaSendCallback` so we can mark the last with with a `@FunctionalInterface`, too, and use it here (or anywhere in the end-user code) as a lambda.

Reply:
I don't think it makes sense; in most real applications you would want to always implement both methods.